File: roboplan_ros_visualization/roboplan_ros_visualization/roboplan_visualizer.py
#!/usr/bin/env python3

import numpy as np
import pinocchio as pin
from typing import Optional

from rclpy.node import Node
from rclpy.qos import (
    QoSProfile,
    QoSReliabilityPolicy,
    QoSHistoryPolicy,
    QoSDurabilityPolicy,
)
from visualization_msgs.msg import Marker, MarkerArray
from std_msgs.msg import ColorRGBA
from sensor_msgs.msg import JointState

from roboplan import Scene
from roboplan_ros_py.type_conversions import se3_to_pose


class RoboplanVisualizer:
    """
    Utility for visualizing a robot configuration using markers in RViz.
    Does not use TF or robot_state_publisher - computes forward kinematics
    directly from the Scene and publishes visualization markers.
    """

    def __init__(
        self,
        node: Node,
        scene: Scene,
        urdf_xml: str,
        package_paths: list = [],
        color: Optional[ColorRGBA] = None,
        update_rate: Optional[float] = None,
        namespace: str = "/roboplan",
    ):
        """
        Initialize the robot visualizer.

        Args:
            node: ROS node instance.
            scene: A fully configured RoboPlan scene.
            urdf_xml: URDF XML string (already processed with xacro if needed).
            package_paths: List of package paths for mesh resolution.
            namespace: Namespace for the marker publisher and joint state subscriber.
            color: Default color for the robot visualization, if set will override defaults.
            update_rate: If provided, subscribes to joint_states and auto-updates.
                         If None, call visualize_configuration() manually.
        """
        self.node = node
        self.scene = scene
        self.namespace = namespace
        self.color = color

        # Build visual geometry model from URDF. I think if/when the scene model has python bindings
        # this will be unnecessary.
        self._build_visual_model(urdf_xml, package_paths or [])

        # Create marker publisher
        qos = QoSProfile(
            depth=1,
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            durability=QoSDurabilityPolicy.VOLATILE,
        )
        self.marker_pub = node.create_publisher(
            MarkerArray, f"{namespace}/markers", qos
        )

        # If update rate provided, subscribe to joint states and update at a fixed rate
        self._joint_state_sub = None
        self._timer = None
        self._latest_joint_positions = None
        if update_rate is not None:
            self._joint_state_sub = node.create_subscription(
                JointState,
                f"{namespace}/joint_states",
                self._joint_state_callback,
                qos,
            )
            self._timer = node.create_timer(1.0 / update_rate, self._timer_callback)

        self.node.get_logger().info(
            f"Constructed RoboplanVisualizer in namespace: {namespace}"
        )

    def _build_visual_model(self, urdf_xml: str, package_paths: list):
        """
        Build visual geometry model from URDF using Pinocchio.

        Like many things this could be removed when Pinocchio has python bindings.
        """
        self.pin_model = pin.buildModelFromXML(urdf_xml)
        self.visual_model = pin.buildGeomFromUrdfString(
            self.pin_model,
            urdf_xml,
            pin.GeometryType.VISUAL,
            package_dirs=package_paths,
        )
        self.visual_data = self.visual_model.createData()

    def _joint_state_callback(self, msg: JointState):
        self._latest_joint_positions = np.array(msg.position)

    def _timer_callback(self):
        if self._latest_joint_positions is not None:
            self.visualize_configuration(self._latest_joint_positions)

    def visualize_configuration(self, q: np.ndarray):
        """
        Visualize a robot configuration using markers.

        Args:
            q: Joint positions (full robot state)
        """
        marker_array = MarkerArray()

        # Update geometry placements based on joint configuration
        pin.updateGeometryPlacements(
            self.pin_model,
            pin.Data(self.pin_model),
            self.visual_model,
            self.visual_data,
            q,
        )
        for idx, geom_obj in enumerate(self.visual_model.geometryObjects):
            try:
                placement = self.visual_data.oMg[idx]
                marker = self._create_geometry_marker(idx, geom_obj, placement)
                if marker:
                    marker_array.markers.append(marker)

            except Exception as e:
                self.node.get_logger().debug(
                    f"Could not create marker for geometry {geom_obj.name}: {e}"
                )
                continue

        if marker_array.markers:
            self.marker_pub.publish(marker_array)

    def _create_geometry_marker(
        self, marker_id: int, geom_obj, placement
    ) -> Optional[Marker]:
        """
        Create a marker for a Pinocchio geometry object.

        Args:
            marker_id: Marker ID
            geom_obj: Pinocchio GeometryObject
            placement: SE3 placement of the geometry

        Returns:
            Marker or None
        """
        marker = Marker()
        marker.header.frame_id = "world"  # or your base frame
        marker.header.stamp = self.node.get_clock().now().to_msg()
        marker.ns = f"{self.namespace}/{geom_obj.name}"
        marker.id = marker_id
        marker.action = Marker.ADD

        # Convert SE3 placement to pose
        marker.pose = se3_to_pose(placement.homogeneous)

        # Handle different geometry types
        # TODO: Something more intelligent with pin's types?
        geometry = geom_obj.geometry
        if isinstance(geometry, pin.hppfcl.Box):
            marker.type = Marker.CUBE
            marker.scale.x = float(geometry.halfSide[0] * 2)
            marker.scale.y = float(geometry.halfSide[1] * 2)
            marker.scale.z = float(geometry.halfSide[2] * 2)

        elif isinstance(geometry, pin.hppfcl.Sphere):
            marker.type = Marker.SPHERE
            r = float(geometry.radius * 2)
            marker.scale.x = marker.scale.y = marker.scale.z = r

        elif isinstance(geometry, pin.hppfcl.Cylinder):
            marker.type = Marker.CYLINDER
            marker.scale.x = float(geometry.radius * 2)
            marker.scale.y = float(geometry.radius * 2)
            marker.scale.z = float(geometry.halfLength * 2)

        elif isinstance(geometry, (pin.hppfcl.Convex, pin.hppfcl.BVHModelBase)):
            # For mesh geometries
            marker.type = Marker.MESH_RESOURCE

            # Ensure proper URI format because RViz gets angry at absolute paths
            if hasattr(geom_obj, "meshPath") and geom_obj.meshPath:
                mesh_path = geom_obj.meshPath

                if not mesh_path.startswith("package://") and not mesh_path.startswith(
                    "file://"
                ):
                    marker.mesh_resource = f"file://{mesh_path}"
                else:
                    marker.mesh_resource = mesh_path
            else:
                # If there isn't a mesh path we can't do anything?
                return None

            if hasattr(geom_obj, "meshScale"):
                marker.scale.x = float(geom_obj.meshScale[0])
                marker.scale.y = float(geom_obj.meshScale[1])
                marker.scale.z = float(geom_obj.meshScale[2])
            else:
                marker.scale.x = marker.scale.y = marker.scale.z = 1.0

        else:
            # Unknown geometry type, skip
            self.node.get_logger().warn(
                f"Unknown geometry type for {geom_obj.name}: {type(geometry)}"
            )
            return None

        # Use embedded colorings for meshes
        if marker.type == Marker.MESH_RESOURCE:
            marker.mesh_use_embedded_materials = True
        # If the geometry has a color user it
        if hasattr(geom_obj, "meshColor") and len(geom_obj.meshColor) >= 4:
            marker.color.r = float(geom_obj.meshColor[0])
            marker.color.g = float(geom_obj.meshColor[1])
            marker.color.b = float(geom_obj.meshColor[2])
            marker.color.a = float(geom_obj.meshColor[3])
        # Apply colorings if set
        if self.color:
            marker.color = self.color

        return marker

    def clear_markers(self):
        marker_array = MarkerArray()
        marker = Marker()
        marker.action = Marker.DELETEALL
        marker_array.markers.append(marker)
        self.marker_pub.publish(marker_array)

    def set_color(self, color: ColorRGBA):
        self.color = color

File: roboplan_ros_visualization/src/roboplan_visualizer.cpp
#include "roboplan_ros_visualization/roboplan_visualizer.hpp"

#include <pinocchio/algorithm/geometry.hpp>
#include <pinocchio/parsers/urdf.hpp>

#include <hpp/fcl/BVH/BVH_model.h>
#include <hpp/fcl/shape/convex.h>
#include <hpp/fcl/shape/geometric_shapes.h>

#include <roboplan_ros_cpp/type_conversions.hpp>

namespace roboplan_ros_cpp {

RoboplanVisualizer::RoboplanVisualizer(const roboplan::Scene& scene, const std::string& urdf_xml,
                                       const std::vector<std::string>& package_paths,
                                       const std::string& frame_id, const std::string& ns,
                                       const std::optional<std_msgs::msg::ColorRGBA>& color)
    : scene_(scene), frame_id_(frame_id), ns_(ns), color_(color) {
  // Build a visual GeometryModel from the URDF.
  // Scene only holds the collision model, so we need this for rendering.
  const auto& model = scene_.getModel();
  std::istringstream urdf_stream(urdf_xml);
  pinocchio::urdf::buildGeom(model, urdf_stream, pinocchio::VISUAL, visual_model_, package_paths);
}

visualization_msgs::msg::MarkerArray
RoboplanVisualizer::visualize_configuration(const Eigen::VectorXd& q) const {
  visualization_msgs::msg::MarkerArray marker_array;

  const auto& model = scene_.getModel();

  pinocchio::Data data(model);
  pinocchio::GeometryData visual_data(visual_model_);
  pinocchio::updateGeometryPlacements(model, data, visual_model_, visual_data, q);

  for (std::size_t idx = 0; idx < visual_model_.geometryObjects.size(); ++idx) {
    const auto& geom_obj = visual_model_.geometryObjects[idx];
    const pinocchio::SE3& placement = visual_data.oMg[idx];

    auto marker = createGeometryMarker(static_cast<int>(idx), geom_obj, placement);
    if (marker) {
      marker_array.markers.push_back(std::move(*marker));
    }
  }

  return marker_array;
}

visualization_msgs::msg::MarkerArray RoboplanVisualizer::clear_markers() {
  visualization_msgs::msg::MarkerArray ma;
  visualization_msgs::msg::Marker m;
  m.action = visualization_msgs::msg::Marker::DELETEALL;
  ma.markers.push_back(m);
  return ma;
}

void RoboplanVisualizer::set_color(const std_msgs::msg::ColorRGBA& color) { color_ = color; }

void RoboplanVisualizer::clear_color() { color_ = std::nullopt; }

std::optional<visualization_msgs::msg::Marker>
RoboplanVisualizer::createGeometryMarker(int marker_id, const pinocchio::GeometryObject& geom_obj,
                                         const pinocchio::SE3& placement) const {

  // The caller can change as needed
  visualization_msgs::msg::Marker marker;
  marker.header.frame_id = frame_id_;
  marker.ns = ns_ + "/" + geom_obj.name;
  marker.id = marker_id;
  marker.action = visualization_msgs::msg::Marker::ADD;
  marker.pose = roboplan_ros_cpp::se3ToPose(placement.toHomogeneousMatrix());

  // Dispatch on hpp-fcl geometry type.
  const auto* geom = geom_obj.geometry.get();

  if (const auto* cyl = dynamic_cast<const hpp::fcl::Cylinder*>(geom)) {
    marker.type = visualization_msgs::msg::Marker::CYLINDER;
    marker.scale.x = marker.scale.y = cyl->radius * 2.0;
    marker.scale.z = cyl->halfLength * 2.0;

  } else if (const auto* cap = dynamic_cast<const hpp::fcl::Capsule*>(geom)) {
    // Capsules are just be approximated as cylinders for now...
    marker.type = visualization_msgs::msg::Marker::CYLINDER;
    marker.scale.x = marker.scale.y = cap->radius * 2.0;
    marker.scale.z = cap->halfLength * 2.0;

  } else if (const auto* box = dynamic_cast<const hpp::fcl::Box*>(geom)) {
    marker.type = visualization_msgs::msg::Marker::CUBE;
    marker.scale.x = box->halfSide[0] * 2.0;
    marker.scale.y = box->halfSide[1] * 2.0;
    marker.scale.z = box->halfSide[2] * 2.0;

  } else if (const auto* sph = dynamic_cast<const hpp::fcl::Sphere*>(geom)) {
    marker.type = visualization_msgs::msg::Marker::SPHERE;
    const double d = sph->radius * 2.0;
    marker.scale.x = marker.scale.y = marker.scale.z = d;

  } else if (dynamic_cast<const hpp::fcl::Cone*>(geom)) {
    // No cone markers are available so...
    // TODO: Make a cone?
    return std::nullopt;

  } else if (dynamic_cast<const hpp::fcl::ConvexBase*>(geom) ||
             dynamic_cast<const hpp::fcl::BVHModelBase*>(geom)) {
    marker.type = visualization_msgs::msg::Marker::MESH_RESOURCE;

    const std::string& mesh_path = geom_obj.meshPath;
    if (mesh_path.empty()) {
      return std::nullopt;
    }

    // Ensure a proper URI scheme so RViz doesn't complain.
    if (mesh_path.rfind("package://", 0) == 0 || mesh_path.rfind("file://", 0) == 0) {
      marker.mesh_resource = mesh_path;
    } else {
      marker.mesh_resource = "file://" + mesh_path;
    }

    marker.scale.x = geom_obj.meshScale[0];
    marker.scale.y = geom_obj.meshScale[1];
    marker.scale.z = geom_obj.meshScale[2];

  } else {
    // TODO: What should we do with these?
    return std::nullopt;
  }

  // Use embedded materials for meshes, if available
  if (marker.type == visualization_msgs::msg::Marker::MESH_RESOURCE) {
    marker.mesh_use_embedded_materials = true;
  }

  // Apply the geometry object's own colour (meshColor is Eigen::Vector4d)
  if (geom_obj.meshColor.size() >= 4) {
    marker.color.r = static_cast<float>(geom_obj.meshColor[0]);
    marker.color.g = static_cast<float>(geom_obj.meshColor[1]);
    marker.color.b = static_cast<float>(geom_obj.meshColor[2]);
    marker.color.a = static_cast<float>(geom_obj.meshColor[3]);
  }

  // Apply the supplied override, if present
  if (color_) {
    marker.color = *color_;
  }

  return marker;
}

}  // namespace roboplan_ros_cpp
